src/librustc_mir/interpret/memory.rs

```diff
@@ -50,7 +54,9 @@ pub struct Memory<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'a, 'mir, 'tcx>> {
     /// deallocation.  When an allocation is not found here, it is a
     /// static and looked up in the `tcx` for read access.  Writing to
     /// a static creates a copy here, in the machine.
-    alloc_map: FxHashMap<AllocId, (MemoryKind<M::MemoryKinds>, Allocation)>,
+    /// We have this in a RefCell to be able to cache when a new allocation
+    /// for a static has to be created on a read access.
+    alloc_map: MonoHashMap<AllocId, (MemoryKind<M::MemoryKinds>, Allocation<M::PointerTag>)>,
```

Comment:
I would have assumed that the code of the `MonoHashMap` would be isolated in miri and not be part of the miri engine. You can add a new associated type to the `Machine` and a trait that offers the interface required by `Memory`.

Reply:
Uh... you want me to put the entire `MonoHashMap` interface into a trait?

I guess I could do that. Makes me wonder whether I shouldn't abstract `Allocation` as well; miri will need some extra stuff in there (for the per-location stack).